added general recommendations (#3)
ssciwr/guidelines: general/README.md

# General Guidelines

The general guidelines provide an overview of sustainable software development and good scientific practice in research software engineering. They provide a general framework which researchers should adhere to when writing scientific software. However, each programming language introduces its own flavor and so the tools that are recommended can be language-specific. Please check the guidelines for the language that you are using. Currently, [C++](../cpp/README.md) and [python](../python/README.md) are available. Please open an issue for requesting recommendations for further programming languages.

## Version control

> Any software that is being actively developed should be under version control. Version control allows you to keep track of any changes to your software and review the history at any time. It also allows you to track issues and work collaboratively. A version-control system for your software should be the default. Different systems are available and at this point we recommend [github](www.github.com).

Reviewer: I would suggest to distinguish the tool (`git`) from the hosting provider (`GitHub`) here. Also I think it is better to explicitly name GitHub, Bitbucket and Gitlab instead of stating that "different systems are available"

Author: Not sure if we should go too much into detail here? Explaining git will be part of the offered course but I thought to keep it as short as possible here for the main recommendations.

## Documentation

> A good documentation is detrimental for ensuring users know what to do with the code, and applying it to the correct use cases. It will save time in the long run, as code by itself is rarely obvious, and it will help outline dependencies. Depending on the programming language used, we recommend different tools like `doxygen`, `sphinx`, or using markdown/a wiki.

Reviewer 2: ```suggestion
Good documentation is essential for ensuring users know what to do with the code, and are applying it to the correct use cases. It will save time in the long run, as code by itself is rarely obvious, and it will help outline dependencies. Depending on the programming language used, we recommend different tools like `doxygen`, `sphinx`, or using markdown/a wiki.
```

> A good documentation is detrimental for ensuring users know what to do with the code, and applying it to the correct use cases. It will save time in the long run, as code by itself is rarely obvious, and it will help outline dependencies. Depending on the programming language used, we recommend different tools like `doxygen`, `sphinx`, or using markdown/a wiki.

Reviewer: Maybe it is worth mentioning that documentation comes at very different levels (inline, generated class docs, manuals, tutorials etc.) and that all of these parts are important - it is not just about using Sphinx or Doxygen.

Author: Same here, I'm not sure if that level of detail is required here. I would think that is rather part of "teaching" than "recommending".

## Testing

> (Automated) testing ensures that errors are detected early and that results are reproducible. Implementing the tests and keeping them updated requires effort, but will pay off in the long run. Reproducibility is a requirement for good scientific practices. Testing also allows for continuous integration/continuos development (deployment) (CI/CD).

Reviewer 2: ```suggestion
(Automated) testing ensures that errors are detected early and that results are reproducible. Implementing the tests and keeping them updated requires effort, but will pay off in the long run. Reproducibility is a requirement for good scientific practices. Testing also makes refactoring easier, and allows for continuous integration/continuous development (deployment) (CI/CD).
```

> (Automated) testing ensures that errors are detected early and that results are reproducible. Implementing the tests and keeping them updated requires effort, but will pay off in the long run. Reproducibility is a requirement for good scientific practices. Testing also allows for continuous integration/continuos development (deployment) (CI/CD).

Reviewer: I am confused by the "Continuous Development" part. CD stands for "continuous deployment", but the term is only to a lesser extent relevant for scientific software IMO. Might as well just leave it at "CI"

> (Automated) testing ensures that errors are detected early and that results are reproducible. Implementing the tests and keeping them updated requires effort, but will pay off in the long run. Reproducibility is a requirement for good scientific practices. Testing also allows for continuous integration/continuos development (deployment) (CI/CD).

Reviewer: It might be worth to include TDD even in this short paragraph.

Reviewer: Maybe also add a short summary of CI providers just like you als mentioned Github in the VC paragraph?

## Refactoring

> Refactoring means that previously implemented routines or legacy code is updated and improved as new parts are added to a software. This allows for a flexible adaptation of the software and ensures that the dependencies are up-to-date. "Always leave code better than you found it" ([source](https://biratkirat.medium.com/step-8-the-boy-scout-rule-robert-c-martin-uncle-bob-9ac839778385)). Make use of a good coding editor and linting to keep your source code clean, and use profilers to check the computational bottlenecks. For example, `VS Code` together with linting extensions and `valgrind` for profiling.

Reviewer 2: ```suggestion
Refactoring means that previously implemented routines or legacy code is updated and improved as new parts are added to a software. This allows for a flexible adaptation of the software and ensures that the dependencies are up-to-date. "Always leave code better than you found it" ([source](https://biratkirat.medium.com/step-8-the-boy-scout-rule-robert-c-martin-uncle-bob-9ac839778385)). Make use of a good coding editor (such as `VS Code`) and linting to keep your source code clean, and use profilers to check the computational bottlenecks.
```
I would leave tools like valgrind to the language-specific guides

Reviewer: ```suggestion
Refactoring means that previously implemented functionality or legacy code is updated and improved as new parts are added to a software. This allows for a flexible adaptation of the software and ensures that the dependencies are up-to-date. "Always leave code better than you found it" ([source](https://biratkirat.medium.com/step-8-the-boy-scout-rule-robert-c-martin-uncle-bob-9ac839778385)). Make use of a good coding editor and linting to keep your source code clean, and use profilers to check the computational bottlenecks. For example, `VS Code` together with linting extensions and `valgrind` for profiling.
```
`routines` sounds awkward outside of Fortran

## Code review

> Code review - involving your collaborators in any changes you made to the code and vice versa - increases the chance of errors being detected early. Furthermore, it also aids knowledge transfer and keeps you up to date what features are being implemented. Code review is easy to incorporate in your work flow through `github pull requests`.

Reviewer 2: ```suggestion
Code review - involving your collaborators in any changes you made to the code and vice versa - increases the chance of errors being detected early. Furthermore, it also aids knowledge transfer and keeps you up to date on what features are being implemented. Code review is easy to incorporate in your work flow through `github pull requests`.
```
